crawler.rb

```diff
@@ -1,39 +1,63 @@
+require 'pry'
+require 'set'
+require 'JSON'
 require 'nokogiri'
 require 'open-uri'
-require 'JSON'
-require 'pry'
 require_relative './document'
-require_relative './document_collection'
 require_relative './url_fetcher'
 
 class Crawler
-	def initialize(document_collection, url_collection)
-		@document_collection = document_collection
-		@url_collection = url_collection
-	end
-
-	def done_crawling?
-		@url_collection.empty?
-	end
-
-	def crawl_next_url
-		url = @url_collection.next_url
-		content = UrlFetcher.fetch(url)
-		document = Document.new(url, content)
-		@document_collection.add_document(document)
-		puts document.domain_hrefs.take(5)
-		document.domain_hrefs.each do |href|
-			@url_collection.add_url(href)
-		end
-	end
-
-	def run(times_to_run)
-		times_to_run.times do
-			if done_crawling? 
-				puts "Crawling complete!"
-				exit 0
-			end
-			crawl_next_url
-		end
-	end
+  attr_accessor :url, :href, :db
+  def initialize(db)
+    @database = db 
+  end
+
+  def done_crawling?
+    @database.execute( "SELECT count(*) FROM URLS_to_crawl WHERE state = 'uncrawled'" )[0][0] == 0
+  end
+
+  def crawl_next_url
+    begin
+      url = @database.execute( "SELECT url FROM URLS_to_crawl WHERE state = 'uncrawled' LIMIT 1" )[0][0]
+      content = UrlFetcher.fetch(url)
+      document = Document.new(url, content)
+      write_documents_to_db(document)
+      document.domain_hrefs.each do |href|
+        write_urls_to_db(href)
+      end
+      update_state(url)
+    rescue => e
+      puts "#{e} #{url}"
+      update_state_for_error(url)
+    end
+  end
+
+  def update_state(url)
+    @database.execute( "UPDATE URLS_to_crawl SET state = 'crawled' WHERE url = ?", [url] )
+  end
+
+  def update_state_for_error(url)
+    @database.execute( "UPDATE URLS_to_crawl SET state = 'error' WHERE url = ?", [url] )
+  end
+
+  def write_urls_to_db(url)
+    @database.execute("INSERT OR IGNORE INTO URLS_to_crawl (url, state) 
+                VALUES (?, 'uncrawled')", ['http://dragonage.wikia.com' + url])
+  end  
+
+  def write_documents_to_db(document)
```

Comment:
This method is grabbing the next url, fetching the content, and inserting the content into the db. It might make sense to rename this method to something more appropriate.